engines/lab/anim.cpp

```diff
@@ -0,0 +1,340 @@
+ *
+ * Copyright (c) 1993 Terra Nova Development
+ * Copyright (c) 2004 The Wyrmkeep Entertainment Co.
+ *
+ */
+
+#include "lab/lab.h"
+
+#include "lab/anim.h"
+#include "lab/dispman.h"
+#include "lab/eventman.h"
+#include "lab/music.h"
+#include "lab/utils.h"
+
+namespace Lab {
+
+Anim::Anim(LabEngine *vm) : _vm(vm) {
+	_header = 0;
+	_curBit = 0;
+	_numChunks = 1;
+	_headerdata._width = 0;
+	_headerdata._height = 0;
+	_headerdata._fps = 0;
+	_headerdata._flags = 0;
+	_delayMicros = 0;
+	_continuous = false;
+	_isPlaying = false;
+	_isAnim = false;
+	_isPal = false;
+	_noPalChange = false;
+	_donePal = false;
+	_frameNum = 0;
+	_playOnce = false;
+	_diffFile = nullptr;
+	_diffFileStart = 0;
+	_size = 0;
+	_rawDiffBM._bytesPerRow = 0;
+	_rawDiffBM._drawOnScreen = false;
+	for (int i = 0; i < 16; i++)
+		_rawDiffBM._planes[i] = nullptr;
+	_waitForEffect = false;
+	_stopPlayingEnd = false;
+	_sampleSpeed = 0;
+	_doBlack = false;
+	_diffWidth = 0;
+	_diffHeight = 0;
+	DrawBitMap = _vm->_graphics->_dispBitMap;
+
+	for (int i = 0; i < 3 * 256; i++)
+		_diffPalette[i] = 0;
+}
+
+void Anim::diffNextFrame(bool onlyDiffData) {
+	if (_header == 65535)
+		// Already done.
+		return;
+
+	if (_vm->_graphics->_dispBitMap->_drawOnScreen)
+		_vm->_graphics->_dispBitMap->_planes[0] = _vm->_graphics->getCurrentDrawingBuffer();
+
+	_vm->_graphics->_dispBitMap->_planes[1] = _vm->_graphics->_dispBitMap->_planes[0] + 0x10000;
+	_vm->_graphics->_dispBitMap->_planes[2] = _vm->_graphics->_dispBitMap->_planes[1] + 0x10000;
+	_vm->_graphics->_dispBitMap->_planes[3] = _vm->_graphics->_dispBitMap->_planes[2] + 0x10000;
+	_vm->_graphics->_dispBitMap->_planes[4] = _vm->_graphics->_dispBitMap->_planes[3] + 0x10000;
+
+	_vm->_event->mouseHide();
+
+	while (1) {
+		if (_curBit >= _numChunks) {
+			_vm->_event->mouseShow();
+
+			if (!onlyDiffData) {
+				if (_headerdata._fps) {
+					uint32 targetMillis = g_system->getMillis() + _delayMicros;
+					while (g_system->getMillis() < targetMillis)
+						g_system->delayMillis(10);
+				}
+
+				if (_isPal && !_noPalChange) {
+					_vm->_graphics->setPalette(_diffPalette, 256);
+					_isPal = false;
+				}
+
+				_donePal = true;
+			}
+
+			if (_isPal && !_noPalChange && !onlyDiffData && !_donePal) {
+				_vm->_graphics->setPalette(_diffPalette, 256);
+				_isPal = false;
+			}
+
+			_donePal = false;
+
+			_frameNum++;
+
+			if ((_frameNum == 1) && (_continuous || !_playOnce))
+				_diffFileStart = _diffFile->pos();
+
+			_isAnim = (_frameNum >= 3) && (!_playOnce);
+			_curBit = 0;
+
+			if (_vm->_graphics->_dispBitMap->_drawOnScreen)
+				_vm->_graphics->screenUpdate();
+
+			// done with the next frame.
+			return;
+		}
+
+		_vm->_music->updateMusic();
+		_header = _diffFile->readUint32LE();
+		_size = _diffFile->readUint32LE();
+
+		uint32 curPos = 0;
+
+		switch (_header) {
+		case 8:
+			_diffFile->read(_diffPalette, _size);
+			_isPal = true;
+			break;
+
+		case 10:
+			if (onlyDiffData)
+				warning("Boom");
```

Comment:
Is this something which should never happen? Or why is there such a (to me rather meaningless) warning in this case?